graders/c/cgrader/cgrader.py

```diff
@@ -367,68 +392,70 @@ def test_compile_file(
             field=field,
         )
 
-    def change_mode(self, file, mode="744", change_parent=True):
+    def change_mode(
+        self, file: str, mode: str = "744", change_parent: bool = True
+    ) -> None:
         file = os.path.abspath(file)
         self.run_command(["chmod", mode, file], sandboxed=False)
         parent = os.path.dirname(file)
         # Ensure that all users can resolve the path name
         if change_parent and parent and not os.path.samefile(file, parent):
             self.change_mode(parent, "a+x")
 
-    def test_send_in_check_out(self, *args, **kwargs):
+    def test_send_in_check_out(self, *args, **kwargs) -> TestResult:
         """Old deprecated function name,
         retained for compatibility reasons."""
         return self.test_run(*args, **kwargs)
 
     def test_run(
         self,
-        command,
-        input=None,
-        exp_output=None,
+        command: str | Iterable[str],
+        input: str | None = None,
+        exp_output: str | Iterable[str] | None = None,
         must_match_all_outputs: OutputMatchingOption | bool = "any",
-        reject_output=None,
-        field=None,
-        ignore_case=True,
-        timeout=1,
-        size_limit=10240,
-        ignore_consec_spaces=True,
-        args=None,
-        name=None,
-        msg=None,
-        max_points=1,
-        highlight_matches=False,
-    ):
+        reject_output: str | Iterable[str] | None = None,
+        field: str | None = None,
+        ignore_case: bool = True,
+        timeout: float = 1,
+        size_limit: int = 10240,
+        ignore_consec_spaces: bool = True,
+        args: str | Iterable[str] | None = None,
+        name: str | None = None,
+        msg: str | None = None,
+        max_points: float = 1,
+        highlight_matches: bool = False,
+    ) -> TestResult:
         if args is not None:
-            if not isinstance(args, list):
+            if isinstance(args, str):
                 args = [args]
             args = list(map(str, args))
```

Comment:
I checked some older uses of this autograder, and it seems there are cases where this function is called with `args` set to an integer value instead of a string or list, which is where this map to str comes into play. I believe this would cause the test two lines above to change behaviour. This shouldn't affect other parameters, only args, but it may be a good idea to use something like `isinstance(args, str) or not isinstance(args, Iterable)` (we'd need both since str is technically an iterable). What do you think?

Reply:
That seems like a good workaround.

Reply:
I think I'd actually prefer `isinstance(args, str | int)`.

Reply:
Just to avoid problems if anyone that I'm not aware of uses floats instead of ints, can you replace it with float? Should be equivalent.

Reply:
Checked in [b8b80ef](https://github.com/PrairieLearn/PrairieLearn/pull/11119/commits/b8b80efd3d36c49f3498856c10635d10ba8f4481).